src/graphql.tsx

```diff
@@ -577,7 +577,7 @@ export default function graphql(
         const clientProps = this.calculateResultProps(data);
         const mergedPropsAndData = assign({}, props, clientProps);
 
-        if (!shouldRerender && renderedElement) {
+        if (!shouldRerender && renderedElement && renderedElement.type === WrappedComponent) {
```

Comment:
Hi @calebmer! I've diagnosed and debugged this with @brunoabreu.
Basically, when a hot module replacement happens, the `render` method of the `GraphQL` is called again and has the correct, new implementation of `WrappedComponent` in scope. 

However, because the wrapped element was already rendered, this `if` is true and we don't get to re-create the element using the new `WrappedComponent` implementation.

We initially checked it works by manually doing `renderedElement = createElement...` and the new component would be rendered.

I believe the intention here is to avoid re-rendering the wrapped component if the data didn't change (which is great), but the original implementation didn't consider that the Component _itself_ might have changed (as is desired during HMR).

This basically checks that the renderedElement is up-to-date with the implementation. If the WrappedComponent didn't change (as it wouldn't without HMR), then it's referentially equal to `renderedElement.type`.

Or maybe we're mistaken, but this seems to be the case from what we've gathered. 😅